// @genkit-ai/tools-common/lib/cjs/src/utils/prompt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromMessages = fromMessages;
const yaml_1 = require("yaml");
function fromMessages(frontmatter, messages) {
    const cleanFrontmatter = cleanupFrontmatter(frontmatter);
    const { rendered: renderedMessages, anyOmitted } = renderMessages(messages);
    const header = `---
${(0, yaml_1.stringify)(cleanFrontmatter, {
        collectionStyle: 'block',
        aliasDuplicateObjects: false,
    }).trim()}
---`;
    if (anyOmitted) {
        return (`${header}

{{! Some advanced message types, such as tool requests/responses, have been omitted from the history. See comments inline for more details. }}

${renderedMessages}`.trimEnd() + '\n');
    }
    return (`${header}

${renderedMessages}`.trimEnd() + '\n');
}
function renderMessages(messages) {
    let anyOmitted = false;
    let rendered = '';
    messages.forEach((message) => {
        const hasToolRequest = message.content.some((p) => 'toolRequest' in p);
        const hasToolResponse = message.content.some((p) => 'toolResponse' in p);
        const hasSupportedPart = message.content.length === 0 ||
            message.content.some((p) => 'text' in p || 'media' in p);
        const hasUnsupportedPart = message.content.some((p) => !('text' in p) && !('media' in p));
        if (hasToolRequest || hasToolResponse || !hasSupportedPart) {
            anyOmitted = true;
            let reason = 'unsupported content';
            if (hasToolRequest) {
                reason = 'toolRequest';
            }
            else if (hasToolResponse) {
                reason = 'toolResponse';
            }
            rendered += `{{! message with role "${message.role}" omitted (${reason}). }}\n\n`;
        }
        else {
            if (hasUnsupportedPart) {
                anyOmitted = true;
            }
            rendered += `{{role "${message.role}"}}\n`;
            rendered += message.content.map(partToString).join('');
            rendered += '\n\n';
        }
    });
    return { rendered, anyOmitted };
}
function cleanupFrontmatter(frontmatter) {
    return recursiveCleanup(frontmatter) || {};
}
function recursiveCleanup(val) {
    if (Array.isArray(val)) {
        const cleaned = val
            .map(recursiveCleanup)
            .filter((v) => v !== undefined && v !== null);
        return cleaned.length > 0 ? cleaned : undefined;
    }
    if (val !== null && typeof val === 'object' && !(val instanceof Date)) {
        const cleaned = {};
        let hasProps = false;
        for (const key in val) {
            const v = recursiveCleanup(val[key]);
            if (v !== undefined && v !== null) {
                cleaned[key] = v;
                hasProps = true;
            }
        }
        return hasProps ? cleaned : undefined;
    }
    return val === null || val === undefined ? undefined : val;
}
function partToString(part) {
    if ('text' in part && part.text !== undefined) {
        return part.text;
    }
    else if ('media' in part && part.media !== undefined) {
        return `{{media url:${part.media.url}}}`;
    }
    const type = Object.keys(part).find((k) => k !== 'metadata' && part[k] !== undefined) || 'unknown';
    return `{{! ${type} part omitted }}`;
}
